Use the MessageBroker to communicate Properties of Services.

ehpor opened the pull request:
This PR makes all Properties of Services be communicated via a message broker. The data inside the messages are still protobuf buffers.

All properties are now stored in shared memory, meaning that there is no time penalty for checking properties anymore. No longer the 1-2ms latency but instant, and at no cost to the service. This allows for example for updating exposure time in the GUI in real time, and the same for the control viewers for HiCAT.

> [!CAUTION]
> This PR branches off from #339 and requires that PR to be merged before this one.

> [!NOTE]
> While this PR may seem small in number of lines of code, it has big impacts on the underlying communication protocols and requires _everyone_ to be on board. I'm requesting review from at least @RemiSoummer and @ivalaginja and will not merge before both have approved. I may include @lanemeier7 in that list as well, depending on his judgement, as stakeholder for GSFC.

Sidenote: I love PRs that increase functionality while reducing the number of lines of code.

## Testing
- [x] Run HiCAT on this PR.
- [x] Run SEAL on this PR.
- [x]  Run THD on this PR. 
- [x] Figure out if the Zernike WFS viewer on HiCAT is broken just with this PR or also on develop.

ehpor committed:
Use the message broker for property values.

ehpor committed:
Get and set property values via the message broker.

ehpor committed:
Move property handling to the Service and the message broker.

ehpor committed:
Have PublishMessage(), PublishData() and PublishArray() return the sent message.

ehpor committed:
Remove stream-backed properties.

I'm keeping the interface the same for now for backwards compatibility.

ehpor committed:
Add == and != operators.

ehpor committed:
Get and publish all properties after Open() but before Main().

ehpor committed:
Add missing include.

ehpor commented:
Looks beautiful though:
<img width="981" alt="image" src="https://github.com/user-attachments/assets/9338626f-a7e9-4e98-999e-6f2907c60ca9" />

ehpor committed:
Add correct trace id to error messages.

ehpor commented:
The only test that is failing is the reading of data-stream-backed properties with the wrong type. I'm going to delete that test, since datastream-backed properties is now deprecated and so the test is irrelevant.

One thing I noticed: upon exiting the pytest, it takes almost a minute to exit. This looks like a timeout of some kind, something that isn't respecting the shutdown signal for the testbed. I'm not sure what it is. Running a `hicat start server` doesn't have the same behaviour. I'll have to look into this more before I feel fine merging this PR.

ehpor committed:
Remove datastream-backed properties since these are now deprecated.

ehpor commented:
Partial log output from running a `pytest`:

```
[testbed] Request received: stop_service
[testbed] Trying to stop service "dummy_service".
[testbed] Creating new socket.
[dummy_service] Request received: shut_down
[dummy_service] Sent reply: OK
[testbed] Sent reply: OK
[testbed] Request received: shut_down
[testbed] Sent reply: OK
[dummy_service] Server has shut down.
[testbed] Shutting down all running services.
[testbed] Creating new socket.
[dummy_dm_service] Request received: shut_down
[dummy_dm_service] Sent reply: OK
[testbed] Creating new socket.
[dummy_dm_service] Server has shut down.
[dummy_service] Service main has ended.
[dummy_service] Closing service.
[dummy_service] Service was closed.
[dummy_dm_service] Service main has ended.
[dummy_dm_service] Closing service.
[dummy_dm_service] Service was closed.
[testbed] The server took too long to respond to our request.
[testbed] Server has shut down.
Process Process-1:
Traceback (most recent call last):
  File "/Users/epor/Github/catkit2/catkit2/testbed/testbed.py", line 170, in stop
    client.make_request('shut_down', request.SerializeToString())
RuntimeError: The server did not respond in time. Is it running?

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/epor/anaconda3/envs/catkit2_test3/lib/python3.11/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/Users/epor/anaconda3/envs/catkit2_test3/lib/python3.11/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/Users/epor/Github/catkit2/tests/conftest.py", line 28, in run_testbed
    testbed.run()
  File "/Users/epor/Github/catkit2/catkit2/testbed/testbed.py", line 394, in run
    self.shut_down_all_services()
  File "/Users/epor/Github/catkit2/catkit2/testbed/testbed.py", line 839, in shut_down_all_services
    shutdown_state[service_id] = self._shut_down_service_with_leniency(service_id, 60, state, t)
                                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Users/epor/Github/catkit2/catkit2/testbed/testbed.py", line 874, in _shut_down_service_with_leniency
    self.services[service_id].stop()
  File "/Users/epor/Github/catkit2/catkit2/testbed/testbed.py", line 172, in stop
    raise RuntimeError("Something went wrong while stopping service.") from e
RuntimeError: Something went wrong while stopping service.
```

ehpor commented:
This appears to be a race condition between the `testbed.stop_service()` call from the dummy_service fixture and the testbed itself shutting down and issuing a shutdown command to the dummy service itself. The second command will of course fail because the dummy_service has already shut down after the first shutdown command. Not sure if that's related to this PR.

ehpor commented:
I'm gonna ignore this race condition. It seems unrelated to the changes in this PR. Probably some of the timings have changed in this PR that weren't the case before. Maybe related to the new thread that is monitoring messages for setting properties that wasn't there before. In any case, this bug isn't there for normal operations, only for the test. And even in the test, sometimes it isn't there.

lanemeier7 commented:
On my Windows / Python 3.8.20 build I am getting the following errors:

``` Pytest
FAILED tests/test_dm_commands.py::test_flatten_single_channel - AttributeError: The server did not respond in time. Is it running?
FAILED tests/test_dm_commands.py::test_flatten_multiple_channels - AttributeError: The server did not respond in time. Is it running?
ERROR tests/test_service.py::test_service_property - RuntimeError: Unable to start service.
ERROR tests/test_service.py::test_service_property_readonly - RuntimeError: Unable to start service.
ERROR tests/test_service.py::test_service_command - RuntimeError: Unable to start service.
ERROR tests/test_service.py::test_service_datastream - RuntimeError: Unable to start service.
================================================================================ 2 failed, 118 passed, 2 skipped, 7 warnings, 4 errors in 253.53s (0:04:13)
```

I checked out the most recent on `develop` and its working fine after installing cfitsio with conda. But when I grab this branch it fails the pytest. Is there something new I need to install?

ehpor commented:
Well, you need to recompile, and there has been an added dependency a few weeks ago.

Do `pytest -s` to see the testbed log output to figure out what is going wrong. @lanemeier7

lanemeier7 commented:
> Well, you need to recompile, and there has been an added dependency a few weeks ago.
> 
> Do `pytest -s` to see the testbed log output to figure out what is going wrong. @lanemeier7

Ok - I think this is an issue with the EntryPoints on Python 3.8.20 - this should be fixed when this branch gets updated from develop.
```
Traceback (most recent call last):
  File "C:\miniforge\envs\catkit2\lib\multiprocessing\process.py", line 315, in _bootstrap
    self.run()
  File "C:\miniforge\envs\catkit2\lib\multiprocessing\process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "C:\HWO\catkit2\tests\conftest.py", line 21, in run_testbed
    testbed = Testbed(port, is_simulated, config)
  File "C:\HWO\catkit2\catkit2\testbed\testbed.py", line 322, in __init__
    module = entry_point.module
AttributeError: 'EntryPoint' object has no attribute 'module'
```

ehpor commented:
@lanemeier7 Rebased onto develop so you can check now.

lanemeier7 reviewed (commented):
I took a look and this doesn't break my setup. However, I'd wait for @RemiSoummer & @ivalaginja to chime in.

Thanks for including me!

lanemeier7 commented:
> @lanemeier7 Rebased onto develop so you can check now.

Yep that fixed it for me

seal-computer committed:
Remove obsolete protobuf message attributes.

Yes, I know you're not supposed to delete them, but we're not worried
about backwards compatibility with these messages.

seal-computer committed:
Remove unused attribute.

ivalaginja commented:
@ehpor sorry to require extra guidance, what is it I am looking at here? Your PR description is rather clear - does this mean I want to test some of the service properties that are now running through the message proper, like the camera exposure time you mentioned? And check them through mtop? Still having some issues on that over in #339 so I will need some help. Thanks!

ehpor commented:
My PR description being rather clear is a great compliment. Usually it's the other way around.

And yes, after this PR, all properties are stored on the message broker. From a Python side, nothing changes other than getters being 1000x faster than before (1ms to 1us), which was before only for the data stream backed properties. Furthermore, you can now "snoop" on the communication between services and experiments/other services.

steigersg commented:
With reference to the Zernike WFS viewer on HiCAT:  on the simulator it opens fine and runs on the most up-to-date version of develop. It is also opens without issue on hardware which is not up-to-date with develop (we're on commit a7dbedf from 27 May). If there is a particular aspect of the viewer that was breaking you want me to check, or if you need a check of the latest version of develop on hardware, let me know.

ehpor commented:
@steigersg Thanks for the confirmation. The viewer was broken when I opened it on my machine. But I didn't try it on develop yet, and I haven't ran the GUI in a long time (almost a year now) on my machine, so it could be just something in my environment that I'm missing. I'll check.

There's no reason why only the Zernike viewer is broken and not anything else, so I'm assuming it's an issue on my end.

steigersg commented:
Confirmed Zernike WFS viewer works on up-to-date develop on HiCAT PC

RemiSoummer commented:
After discussion with @ehpor @steigersg and @ivalaginja we all are on the same page with the scope of the PR

ehpor committed:
Add try-catch to catch any exceptions when getting a property.

ehpor commented:
@steigersg I added a try-catch to catch any exceptions that occur during getting of the `device_name` property, log those as errors, then continue with the other properties. You should try testing on hardware again.

RemiSoummer committed:
adding the missing attribute

RemiSoummer commented:
I tested on hardware and still saw an issue with the missing attribute for `device_name` so I fixed it.  before/ after upon starting the zwo camera. 
<img width="876" alt="Screenshot 2025-07-02 at 10 46 58 AM" src="https://github.com/user-attachments/assets/2ead1af3-c8ec-4c30-8ead-3d7d84879012" />
<img width="643" alt="Screenshot 2025-07-02 at 11 11 10 AM" src="https://github.com/user-attachments/assets/c939002d-b84f-4826-b06f-f351dee6e09a" />

@ehpor fyi since I pushed to your PR.

RemiSoummer commented:
I tested a bunch of stuff on hardware, both types of cameras ZWO and FLIR,  flip mounts, motors, etc.  I'm not seeing any errors in the server and seems that it's creating and handling all the properties for each devices correctly, and I can change cameras properties like exposure time and ROI parameters and everything seems fine.   

Same behavior in simulation my Mac.

RemiSoummer reviewed (approved):
I concur with the scope of the PR after discussion with @ehpor and I have tested successfully now both on sim and on the hicat hardware.   Approving for the HiCAT team.

RemiSoummer commented:
Note : Do not merge until we have approval from @ivalaginja for THD team

RemiSoummer commented:
@lanemeier7 can you confirm that you are okay with the scope of the PR and its implications?

lanemeier7 commented:
yes, I am OK with this PR and implications. Thank you for checking

ehpor commented:
https://github.com/spacetelescope/catkit2/pull/352#issuecomment-3028259192

@RemiSoummer Great! Yeah, when we tested yesterday, because that attribute didn't exist, the service crashed outright and the `close()` wasn't even called. The ZWO camera not having that attribute by mistake made us catch a much bigger bug.

Also, I cannot believe that in almost 4 years, since the start of catkit2, with many people working on multiple testbeds, literally no-one ever called that property 🤣

ivalaginja reviewed (approved):
I am on board with the changes in this PR.

I ran the simulated thd2 server successfully on this branch, together with the GUI main window, service state viewer, and motor state viewer. I was able to normally run the simulated services for a camera, our filter wheel, laser diodes, and three of the motors. Looks like it's ready to go.

ivalaginja commented:
> I tested on hardware and still saw an issue with the missing attribute for `device_name` so I fixed it. before/ after upon starting the zwo camera.

Hey @RemiSoummer, on which branch/where did you fix this? I just saw that this attribute already existed as `self.expected_device_name` 😬  Looks like a dupe. Do we really need `self.device_name` as well?
